core/test/mpi/base/bindings.cpp

```diff
@@ -290,174 +408,714 @@ TYPED_TEST(MpiBindings, CanGetValuesWithFence)
-    ASSERT_EQ(sum, TypeParam{16});
+    auto win = window(data.data(), 4, comm);
+
+    if (my_rank != 0) {
+        win.lock(0);
+        win.get(data.data(), 4, 0, 0, 4);
+        win.unlock(0);
+    }
+
+    auto ref = std::vector<TypeParam>{1, 2, 3, 4};
+    ASSERT_EQ(data, ref);
 }
 
 
-TYPED_TEST(MpiBindings, CanAllReduceValuesInPlace)
+TYPED_TEST(MpiBindings, CanGetValuesWithFence)
 {
+    using window = gko::mpi::window<TypeParam>;
     auto comm = gko::mpi::communicator(MPI_COMM_WORLD);
     auto my_rank = comm.rank();
     auto num_ranks = comm.size();
-    TypeParam data;
+    std::vector<TypeParam> data;
     if (my_rank == 0) {
-        data = 3;
-    } else if (my_rank == 1) {
-        data = 5;
-    } else if (my_rank == 2) {
-        data = 2;
-    } else if (my_rank == 3) {
-        data = 6;
+        data = std::vector<TypeParam>{1, 2, 3, 4};
+    } else {
+        data = std::vector<TypeParam>{0, 0, 0, 0};
     }
-    comm.all_reduce(&data, 1, MPI_SUM);
-    ASSERT_EQ(data, TypeParam{16});
+    auto win = window(data.data(), 4, comm);
+
+    win.fence();
+    if (my_rank != 0) {
+        win.get(data.data(), 4, 0, 0, 4);
+    }
+    win.fence();
+
+    auto ref = std::vector<TypeParam>{1, 2, 3, 4};
+    ASSERT_EQ(data, ref);
 }
 
 
-TYPED_TEST(MpiBindings, CanScatterValues)
+TYPED_TEST(MpiBindings, CanGetAccumulateValuesWithLockAll)
 {
+    using window = gko::mpi::window<TypeParam>;
     auto comm = gko::mpi::communicator(MPI_COMM_WORLD);
     auto my_rank = comm.rank();
     auto num_ranks = comm.size();
-    auto scatter_from_array = gko::Array<TypeParam>{this->ref};
+    std::vector<TypeParam> data;
+    std::vector<TypeParam> target;
+    std::vector<TypeParam> result(4, 0);
     if (my_rank == 0) {
-        scatter_from_array =
-            gko::Array<TypeParam>{this->ref, {2, 3, 1, 3, -1, 0, 3, 1}};
+        data = std::vector<TypeParam>{1, 2, 3, 4};
+        target = std::vector<TypeParam>{1, 2, 3, 4};
+    } else if (my_rank == 1) {
+        data = std::vector<TypeParam>{5, 6, 7, 8};
+        target = std::vector<TypeParam>{5, 6, 7, 8};
+    } else if (my_rank == 2) {
+        data = std::vector<TypeParam>{9, 10, 11, 12};
+        target = std::vector<TypeParam>{9, 10, 11, 12};
+    } else {
+        data = std::vector<TypeParam>{0, 0, 0, 0};
+        target = std::vector<TypeParam>{0, 0, 0, 0};
     }
-    auto scatter_into_array = gko::Array<TypeParam>{this->ref, 2};
-    comm.scatter(scatter_from_array.get_data(), 2,
-                 scatter_into_array.get_data(), 2, 0);
-    auto comp_data = scatter_into_array.get_data();
-    if (my_rank == 0) {
-        ASSERT_EQ(comp_data[0], TypeParam{2});
-        ASSERT_EQ(comp_data[1], TypeParam{3});
 
-    } else if (my_rank == 1) {
-        ASSERT_EQ(comp_data[0], TypeParam{1});
-        ASSERT_EQ(comp_data[1], TypeParam{3});
+    {
+        auto win = window(target.data(), 4, comm);
+
+        if (my_rank == 2) {
+            win.lock_all();
```

Comment:
Do we need locks here? Can we do this test without locks? It might be better in that case - to make this a proper *unit* test.

Reply:
Yes, as it is a one-sided communication, you need either a passive target sync or an active synchronization (fence type) to actually allow the communication.